host/sgx/windows/enter.asm

```diff
@@ -109,71 +84,66 @@ execute_eenter:
     mov rax, ENCLU_EENTER
     ENCLU
 
-    mov ARG1OUT, rdi
-    mov ARG2OUT, rsi
+    ;; Save return values
+    mov ARG1, rdi
+    mov ARG2, rsi
 
-    ;; Restore the saved host context
-    mov rcx, HOST_CONTEXT
-    call oe_restore_host_context
+    ;; Restore flags and floating point state.
+    popfq
+    fxrstor FX_SPACE
+
+    ;; Check if an OCALL needs to be dispatched.
+    mov r10, rdi
+    shr r10, 48
+    cmp r10, OE_OCALL_CODE
+    jne return_from_ecall
 
 dispatch_ocall:
     ;; RAX = __oe_dispatch_ocall(
     ;;     RCX=arg1
     ;;     RDX=arg2
-    ;;     R8=arg1_out
-    ;;     R9=arg2_out
+    ;;     R8=&arg1
+    ;;     R9=&arg2
     ;;     [RSP+32]=TCS,
     ;;     [RSP+40]=ENCLAVE);
     ;;
     ;; Stack should already be 16-byte aligned, so only need
     ;; shadow space (32 bytes) plus stack params size (16 bytes)
     sub rsp, 48
-    mov rcx, ARG1OUT
-    mov rdx, ARG2OUT
-    lea r8, qword ptr ARG1OUT
-    lea r9, qword ptr ARG2OUT
+    mov rcx, rdi
+    mov rdx, rsi
+    lea r8, ARG1
+    lea r9, ARG2
     mov rax, qword ptr TCS
     mov qword ptr [rsp+32], rax
-    mov rax, qword ptr ENCLAVE
-    mov qword ptr [rsp+40], rax
+    mov rax, ENCLAVE
+    mov [rsp+40], rax
     call __oe_dispatch_ocall ;; RAX contains return value
     add rsp, 48
 
-    ;; Restore the stack pointer
-    mov rsp, STACKPTR
-
-    ;; If this was not an OCALL, then return from ECALL.
-    cmp rax, 0
-    jne return_from_ecall
-
-    ;; Stop speculative execution at fallthrough of conditional check
-    lfence
-
-    ;; Prepare to reenter the enclave, calling the entry point.
-    mov rax, ARG1OUT
-    mov ARG1, rax
-    mov rax, ARG2OUT
-    mov ARG2, rax
     jmp execute_eenter
 
 return_from_ecall:
-    ;; Stop speculative execution at target of conditional jump
-    lfence
-
-    ;; Set ARG3 (out)
-    mov r10, ARG1OUT
-    mov rax, qword ptr ARG3_PARAM
-    mov qword ptr [rax], r10
-
-    ;; Set ARG4 (out)
-    mov r10, ARG2OUT
-    mov rax, qword ptr ARG4_PARAM
-    mov qword ptr [rax], r10
-
-    ;; Return parameters space and restore the stack pointer
-    mov rsp, rbp
+    ;; Write results
+    ;; RSI contains arg1, RDI contains arg2
+    mov rbx, ARG3
+    mov [rbx], rdi
+    mov rcx, ARG4
+    mov [rcx], rsi
+
+    ;; Restore callee saved registers
+    fxrstor FX_SPACE
```

Comment:
Again, this seems a little weird because you've separated out `fxrstor` from restoring registers in the context `execute_eenter` but we've effectively double-restored the FX state, once for on return from EENTER for the FX state, and here again for the XMM registers, even though each invocation restores everything. It deserves a comment at least, and from an encapsulation perspective, I'd say it's cleanest to always save and restore the full context as before rather than try to optimize partial contexts for save & restore. 

Alternatively, consider that restoring callee-saved registers could be done once on return from the EENTER instead of at the end of the function: we have to do it insofar as host doesn't trust enclave to preserve them, but within the context of this function, you should be able to definitively say whether any of these were modified (e.g. a call to dispatch should preserve callee-saved registers on your behalf) and we generally shouldn't have to restore them again.